### ISSUE TITLE
Generic type arguments are not encoded properly

### ISSUE THREAD

**Issue by estrizhok** [xunit/xunit#1916]
XUnit encodes all of information about a test in a base64 value. This encoding, however, does not work for a test receiving types of generic arguments as its arguments. To demonstrate the issue we need R# Test Runner as it allows running a single test-case from all test-cases for a given theory.

Steps to reproduce:

1. Create a unit-test project targeting latest .netcoreapp supported by xunit.
2. Create the following test class:
```
using System;
using System.Collections.Generic;
using System.Runtime.Serialization;
using Xunit;

namespace Repro
{
  public class UnitTest1
  {
    [Theory]
    [MemberData(nameof(Data))]
    public void TestReceivingGenericArgumentsAsParameters(Type[] args)
    {
      Assert.All(args, arg => Assert.DoesNotContain(
        arg.GetGenericParameterConstraints(), 
        c => c == typeof(ISerializable)
      ));
    }

    public static IEnumerable<object[]> Data()
    {
      var type = typeof(Component);
      var generic = type.GetMethod("Generic");
      var constrained = type.GetMethod("ConstrainedGeneric");

      yield return new object[] { generic.GetGenericArguments() };
      yield return new object[] { constrained.GetGenericArguments() };
    }
  }

  public class Component
  {
    public void Generic<T>() { }
    public void ConstrainedGeneric<U>() where U : ISerializable { }
  }
}
```
3. Build the projects and all test-cases of `TestReceivingGenericArgumentsAsParameters2` should be discovered by ReSharper.
4. Try running any one of them.
5. The test-case will become inconclusive and the following warning will appear in execution log:
```
[xUnit.net 00:00:00.34] XUnitTestProject2: Catastrophic error during deserialization: System.ArgumentException: String cannot have zero length.
   at System.Reflection.RuntimeAssembly.GetType(RuntimeAssembly assembly, String name, Boolean throwOnError, Boolean ignoreCase, ObjectHandleOnStack type, ObjectHandleOnStack keepAlive)
   at System.Reflection.RuntimeAssembly.GetType(String name, Boolean throwOnError, Boolean ignoreCase)
   at Xunit.Serialization.XunitSerializationInfo.Deserialize(Type type, String serializedValue) in C:\Dev\xunit\xunit\src\common\XunitSerializationInfo.cs:line 175
   at Xunit.Serialization.XunitSerializationInfo.DeserializeTriple(String value) in C:\Dev\xunit\xunit\src\common\XunitSerializationInfo.cs:line 93
   at Xunit.Serialization.XunitSerializationInfo.DeserializeSerializable(Type type, String serializedValue) in C:\Dev\xunit\xunit\src\common\XunitSerializationInfo.cs:line 198
   at Xunit.Serialization.XunitSerializationInfo.Deserialize(Type type, String serializedValue) in C:\Dev\xunit\xunit\src\common\XunitSerializationInfo.cs:line 182
   at Xunit.Serialization.XunitSerializationInfo.DeserializeTriple(String value) in C:\Dev\xunit\xunit\src\common\XunitSerializationInfo.cs:line 93
   at Xunit.Serialization.XunitSerializationInfo.DeserializeSerializable(Type type, String serializedValue) in C:\Dev\xunit\xunit\src\common\XunitSerializationInfo.cs:line 198
   at Xunit.Serialization.XunitSerializationInfo.Deserialize(Type type, String serializedValue) in C:\Dev\xunit\xunit\src\common\XunitSerializationInfo.cs:line 182
   at Xunit.Serialization.XunitSerializationInfo.DeserializeTriple(String value) in C:\Dev\xunit\xunit\src\common\XunitSerializationInfo.cs:line 93
   at Xunit.Serialization.XunitSerializationInfo.DeserializeSerializable(Type type, String serializedValue) in C:\Dev\xunit\xunit\src\common\XunitSerializationInfo.cs:line 198
   at Xunit.Serialization.XunitSerializationInfo.Deserialize(Type type, String serializedValue) in C:\Dev\xunit\xunit\src\common\XunitSerializationInfo.cs:line 110
   at Xunit.Sdk.SerializationHelper.Deserialize[T](String serializedValue) in C:\Dev\xunit\xunit\src\common\SerializationHelper.cs:line 40
   at Xunit.Sdk.XunitTestFrameworkExecutor.Deserialize(String value) in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\Frameworks\XunitTestFrameworkExecutor.cs:line 84
   at System.Linq.Utilities.<>c__DisplayClass2_0`3.<CombineSelectors>b__0(TSource x)
   at System.Linq.Enumerable.SelectListIterator`2.ToList()
   at Xunit.Runner.VisualStudio.VsTestRunner.RunTestsInAssembly(IRunContext runContext, IFrameworkHandle frameworkHandle, LoggerHelper logger, TestPlatformContext testPlatformContext, RunSettings runSettings, IMessageSinkWithTypes reporterMessageHandler, AssemblyRunInfo runInfo) in C:\Dev\xunit\xunit\src\xunit.runner.visualstudio\VsTestRunner.cs:line 562
```

Decoded base64 value shows that information about type argument was omitted (highlighted in bold close to the end of the snippet):
<pre>
TestMethod:Xunit.Sdk.TestMethod, xunit.execution.{Platform}:MethodName:System.String:TestReceivingGenericArgumentsAsParameters
TestClass:Xunit.Sdk.TestClass, xunit.execution.{Platform}:TestCollection:Xunit.Sdk.TestCollection, xunit.execution.{Platform}:DisplayName:System.String:Test collection for Repro.UnitTest1
TestAssembly:Xunit.Sdk.TestAssembly, xunit.execution.{Platform}:AssemblyPath:System.String:Repro.dll
ConfigFileName:System.Object
Version:System.String:1.0.0.0
UniqueID:System.String:00e1bb8e-972e-4248-83c4-28ab15d94afc
DeclarationAssemblyName:System.Object
DeclarationTypeName:System.Object
ClassAssemblyName:System.String:Repro, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null
ClassTypeName:System.String:Repro.UnitTest1
TestMethodArguments:System.Object[]:ElementType:System.String:System.Object
Rank:System.Int32:1
TotalLength:System.Int32:1
Length0:System.Int32:1
LowerBound0:System.Int32:0
Item0:System.Type[]:ElementType:System.String:System.Type
Rank:System.Int32:1
TotalLength:System.Int32:1
Length0:System.Int32:1
LowerBound0:System.Int32:0
<b>Item0:System.Type:, Repro</b>
DefaultMethodDisplay:System.String:ClassAndMethod
DefaultMethodDisplayOptions:System.String:None
Timeout:System.Int32:0
</pre>

**Comment by bradwilson** [comment]
Result from running with the .NET Framework console runner:

![image](https://user-images.githubusercontent.com/16944/55439421-66ffc800-5559-11e9-91cc-79d94333f522.png)

The .NET Framework results are probably what you expected: one passes (T), one fails (U).

Result from running with the .NET Core VSTest runner:

![image](https://user-images.githubusercontent.com/16944/55439262-dde89100-5558-11e9-96c2-f420715c7403.png)

The VSTest results are probably not what you expected, because the argument serialization inherent in VSTest ended up with two identical unique IDs for T and U. While I sort of expected this result, it could be argued that it is a bug (that we might not be able to fix).

The specific R# runner behavior that you're showing as a bug is actually a bug in the R# runner, because it discovered tests it shouldn't be able to run. So there are two bugs here: one ours (the T and U versions have identical IDs), one Resharpers (it shows you tests it can't run).

I will leave this bug open here in the even that maybe we can find/fix our issue, but the R# issue should be reported separately to the R# team.

**Comment by estrizhok** [comment]
I believe you incorrectly identified the root cause. Yes, the IDs are the same. (I actually did not check that, but it is a different problem, and it should not matter in this case. It might have been a single test case, thus having a unique id, the result would have still been the same). But whatever the IDs are, there are two test cases reported by the xUnit adapter, when an IDE executes a discovery for the project. Here I am talking about `dotnet vstest` command. 

Here comes an error.

If the IDE then tries to execute any of those test cases by specifying them explicitly in [TestRunRequestPayload.TestCases](https://github.com/Microsoft/vstest/blob/master/src/Microsoft.TestPlatform.ObjectModel/Client/Payloads/TestRunRequestPayload.cs), the xUnit adapter will fail to deserialize that test case (and abort execution completely, too, even if it is not the only test case in the run).

It is different from running tests using `dotnet test` command in your example, because (as I understand it) the latter discovers and runs in the same process at about the same moment in time, thus eliminating the need to deserialize test cases.

I disagree that it is a bug of R#. All R# does is shows the test cases xUnit adapter reports. But when R# tries to feed the same test cases back to xUnit adapter, telling it to run them, the adapter crushes in deserialization. And it only those types of tests (ones with types of generic arguments passed are arguments) that are causing the problem. All others can be executed just fine.

Hope this helps.

**Comment by bradwilson** [comment]
I have acknowledged that we have a bug, so you don't have to argue that we have a bug. 😉

**Comment by estrizhok** [comment]
Errr.. Yeaaaaah. But we were talking about different bugs. Weren't we?

**Comment by bradwilson** [comment]
Fixed in v2 `2.8.2-pre.16`
Fixed in v3 `0.1.1-pre.485`

https://xunit.net/docs/using-ci-builds

I looked at ways I could serialize these types, but it wasn't really feasible, so the next best thing was to disable serialization for these types. The end result is that your code works now, but you will not see individually runnable theories any more.

![image](https://github.com/xunit/xunit/assets/16944/fa239886-7792-4208-9572-8802df4751b2)

![image](https://github.com/xunit/xunit/assets/16944/36da6d71-ee9d-4c7b-ba3d-677ff2908eff)
